Goal display

Stokalas opened the pull request:
**Goal display:**
- Displays goals dinamically by 1,2 or 3 in a row.
- Remove goals
- When savings are >=100% of required amount for goal ability button to remove goal and it to expenses
- Working progress bars
- Modal to enter new goals
- Base to use any user ID, but 0 left as placeholder.

Stokalas committed:
1 goal display

Stokalas committed:
Dynamic goal display form List

Stokalas committed:
Remove goal from db by id and userId; goal front improvements

Stokalas committed:
UserId in goals optimisation

Stokalas committed:
Add goals

Stokalas committed:
small fixes in add Goal

Stokalas committed:
Ability to add goal to expenses when bought

Stokalas committed:
Fixed 'unnamed' bug when editing entry

Stokalas committed:
Code improvements on Goals.cshtml.cs

Stokalas committed:
space between binded fields

Stokalas committed:
Code optimisation and bug fix

Stokalas committed:
Displaying progress bar for goals; Titles when hovered in /Goals; can add goal to expenses only when saved required amount

Stokalas committed:
Merge branch 'auth' into GoalDisplay

Stokalas committed:
Renamed links to lowercase

Stokalas committed:
Limit goal title to 30 chars; quote when no goals

Stokalas committed:
Changed goal addition from form to modal

Stokalas committed:
removed empty line

ciliakas commented on src/Pages/EditEntry.cshtml.cs:
```diff
@@ -37,37 +37,38 @@ public void OnGet(int id, int entryType)
         /*If cancel pressed return to previous page*/
-        public void OnPostCancel()
+        public IActionResult OnPostCancel()
         {
-            if (EntryTypeInt == 1)
-            {
-                Response.Redirect("/Income");
-                return;
-            }
-            Response.Redirect("/Expenses");
+            return RedirectToPage(EntryTypeInt == 1 ? "/Income" : "/Expenses");
```
Nice improvement to this whole class, the code is a lot more tidy and easier to understand

ciliakas commented on src/Pages/Goals.cshtml:
```diff
@@ -1,36 +1,108 @@
+        @foreach (var item in Model.Goals)
+        {
+            if (num % 3 == 0)
+            {
+                @Html.Raw("<div class=\"row p-0 mx-auto mt-3\">")
+            }
+            <div class="border m-3 bg-primary col-sm">
+                <button asp-page-handler="Delete" asp-route-id="@item.Id" class="btn btn-primary float-right" type="submit" title="Remove Goal"><i class="far fa-trash-alt text-danger "></i></button>
+                <div class="clearfix"></div>
+                <h3 class="text-light problematic-title" title="Title">@item.Title </h3>
+                <div class="progress mt-4">
+                    @{
+                        var percent = decimal.Round(Model.Savings / item.Amount * 100);
+                        percent = percent > 100 ? 100 : percent;
+                    }
+                    <div class="progress-bar bg-success" role="progressbar" style="width: @percent%;" aria-valuenow="@percent" aria-valuemin="0" aria-valuemax="100" title="Progress">@percent%</div>
+                </div>
+                <h5 class="text-light" title="Saved/Required">@Model.Savings/@item.Amount</h5>
+                @if (percent == 100)
+                {
+                    <button asp-page-handler="Purchased" asp-route-id="@item.Id" asp-route-title="@item.Title" asp-route-amount="@item.Amount" class="btn btn-primary" type="submit" title="Add Goal as expense and remove it"><i class="fas fa-shopping-cart fa-2x text-success"></i></button>
+                }
+                else
+                {
+                    <i class="fas fa-piggy-bank fa-2x text-warning" title="Still saving"></i>
+                }
+
+            </div>
+            num++;
+            if (num % 3 == 0)
+            {
+                @Html.Raw("</div>")
+            }
+        }
+    </form>
+    if (num % 3 != 0)
+    {
+        @Html.Raw("</div>")
+    }
+}
```
I really like the system you created for displaying of goals, could you explain a bit how it works?

Stokalas replied:
- Basically it goes through all goals, and when number is divisible by 3 i.e. **0**,3,6... (because max 3 elements in one row) it creates new row:
` @Html.Raw("<div class=\"row p-0 mx-auto mt-3\">")` - just a tag with class that puts items into one line.
- After adding goal, if it's number is divisible by 3 it closes the div, since full row is made (pay attention that closing happens after new item added and incremented, and opening before adding the item:
`if (num % 3 == 0)
            {
                @Html.Raw("</div>")
            }`
- After going through all items if last row was not closed (because it was not filled to have 3 items, it checks that (number is not divisible by 3 and closes last row):
`if (num % 3 != 0)
    {
        @Html.Raw("</div>")
    }`

_This is the logic of displaying all goals "dinamically", if you have questions about other aspects of this code don't hesitate to ask._

ciliakas commented on src/Pages/Goals.cshtml.cs:
```diff
@@ -1,13 +1,77 @@
         {
+            var dataManager = new DataManager();
+            Goals = dataManager.Goals.GoalList;
+            UserId = int.Parse(User.FindFirst(ClaimTypes.Name).Value);
```
Am I understanding correctly that you save and get the user id in the cookies so you don't have to access the database to find it?

Stokalas replied:
Yep, it is saved in these claims, you've asked about in previous pull request.

ciliakas commented on src/Pages/Goals.cshtml.cs:
```diff
@@ -1,13 +1,77 @@
+        {
+            decimal.TryParse(amount, out var parsedAmount);
+            var entry = new Entry(title, parsedAmount, DateTime.Today, recurring:false, importance:1);
+            EntryDbUpdater.Add(entry, 0, EntryType.Expense);
+            DeleteGoalFromDb(id);
```
Later on I'll make a method for this so you don't have to implement this logic in the front end itself :)

Stokalas replied:
I wanted to ask you for that, but now just for testing how it works I've written it in front.

ciliakas reviewed (changes requested):
Good work, I really like the goal display and that we can display as many goals as we want. However a couple of things I think need to be addressed:
- When adding goals, when we leave input fields empty and click add goal, it closes the dialog/form instead of showing what the user inputted wrong ( when you click add goal again, we can see the red error labels)
- This is really minor, but the green color used for the add to expenses button really hurts my eyes, when contrasted with the blue

Stokalas committed:
Alert when wrong goal info added; different coloring for different progress bars; cart color is not that bright

Stokalas commented:
1. I was not able to display Modal on page load after wrong data added. It requires  Javascript etc. And since I'm trying to keep code "C#-arpy" (none JS used by me, except one that comes with bootstrap etc) I did not implement it, just added alert that informs user that something went wrong.
2. Changed color of cart that adds goal to expenses.
3. Progress bar is more colorful than ever: 
   - 0-29% - Red (_Bootstrap: danger_)
   - 30-84% - Yellow  (_Bootstrap: warning_)
   - 85-99% - Electric  (_Bootstrap: info_)
   - 100% - Green (_Bootstrap: success_).

Stokalas committed:
Changed alert message

ciliakas reviewed (approved):
Colors have been adjusted, and if there's no easy fix for the dialog issue then really it's not worth the effort to spend that much time on a small detail. Good job, approving.
